### PR TITLE
devicetree: extern dt device structs

### PR DESCRIPTION
**Pull request by galak**
Add extern of possible device struct based on DT so we don't have to put
a DEVICE_DT_DECLARE everywhere.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

### CODE REVIEW

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
As discussed on Slack, this doesn't belong in devicetree.h. The device layer is higher up.

**Comment by galak** [comment]
Moved to creating a separate file `device_extern.h`

File: scripts/dts/gen_defines.py [hunk 1]
```diff
@@ -110,6 +110,28 @@ def main():
...
+        print("", file=dev_header_file)
+        print("#ifndef _ASMLANGUAGE", file=dev_header_file)
+        print("", file=dev_header_file)
+        print("#ifdef __cplusplus", file=dev_header_file)
+        print('extern "C" {', file=dev_header_file)
+        print("#endif", file=dev_header_file)
+        print("", file=dev_header_file)
+
+        for node in sorted(edt.nodes, key=lambda node: node.dep_ordinal):
+            print(f"extern const struct device __device_DT_{node.z_path_id};", file=dev_header_file)
```
Comment (pabigot) [inline scripts/dts/gen_defines.py:125]: I'd rather not have this hard-code the representation of the external symbol.  Can we move the include of the generated file to the end of `device.h` and have this just use `DT_DEVICE_DECLARE();`?

**Review by pabigot — CHANGES_REQUESTED** [review]
Don't want this hard-coding what a maintained macro generates.

File: scripts/dts/gen_defines.py (hunk 1 above)
Comment (galak) [reply scripts/dts/gen_defines.py:125]: I like that, will rework this to use DT_DEVICE_DECLARE().

File: include/device.h [hunk 2]
```diff
@@ -21,6 +21,7 @@
 
 #include <init.h>
 #include <sys/device_mmio.h>
+#include <device_extern.h>
```
Comment (mbolivar-nordic) [inline include/device.h:24]: Maybe comment that this is a generated file?

File: scripts/dts/gen_defines.py [hunk 3]
```diff
@@ -110,6 +110,28 @@ def main():
         write_chosen(edt)
         write_global_compat_info(edt)
 
+    with open(args.device_header_out, "w", encoding="utf-8") as dev_header_file:
+        print("#ifndef DEVICE_EXTERN_GEN_H", file=dev_header_file)
+        print("#define DEVICE_EXTERN_GEN_H", file=dev_header_file)
+        print("", file=dev_header_file)
+        print("#ifndef _ASMLANGUAGE", file=dev_header_file)
```
Comment (mbolivar-nordic) [inline scripts/dts/gen_defines.py:117]: Not needed since it's in device.h now.

File: scripts/dts/gen_defines.py [hunk 4]
```diff
@@ -110,6 +110,28 @@ def main():
         write_chosen(edt)
         write_global_compat_info(edt)
 
+    with open(args.device_header_out, "w", encoding="utf-8") as dev_header_file:
```
Comment (mbolivar-nordic) [inline scripts/dts/gen_defines.py:113]: Would you mind making a separate function for this?

**Comment by mbolivar-nordic** [comment]
This seems to make DEVICE_DT_DECLARE() more of an implementation detail than API users need to know about, wouldn't you say @pabigot ?

**Comment by pabigot** [comment]
> This seems to make DEVICE_DT_DECLARE() more of an implementation detail than API users need to know about

I suppose so, yes.  In which case, if this gets in, we'll need to update the documentation again.  Or perhaps that could be part of this PR....

**Comment by mbolivar-nordic** [comment]
>  In which case, if this gets in, we'll need to update the documentation again.

I'll handle it this time.

**Comment by mbolivar-nordic** [comment]
I think that means this PR should rename DEVICE_DT_DECLARE to Z_DEVICE_DT_DECLARE and remove all the users in the tree too.

File: scripts/dts/gen_defines.py (hunk 3 above)
Comment (galak) [reply scripts/dts/gen_defines.py]: done.

File: include/device.h (hunk 2 above)
Comment (galak) [reply include/device.h]: Added comment.

File: scripts/dts/gen_defines.py (hunk 4 above)
Comment (galak) [reply scripts/dts/gen_defines.py]: Done.

File: include/device.h [hunk 5]
```diff
@@ -767,6 +767,9 @@ static inline int device_pm_put_sync(const struct device *dev) { return -ENOTSUP
 }
 #endif
 
+/* device_extern is generated base on devicetree nodes */
```
Comment (mbolivar-nordic) [inline include/device.h:732]: "based on" or "from"

**Review by mbolivar-nordic — APPROVED** [review]
The device.h comment has a typo. Looks good otherwise.

File: include/device.h (hunk 5 above)
Comment (galak) [reply include/device.h]: fixed.

Comment (pabigot) [reply include/device.h]: s/base/based/

**Review by pabigot — APPROVED** [review]

**Pull request merged**